### ISSUE TITLE
PacBio Consensus sequencing using UMI-barcode + DADA2

### ISSUE THREAD

**Issue by Ellie-BM** [benjjneb/dada2#1054]
Hi Ben,

Is it possible to process the HIV sequences obtained from PacBio CCS mode where RT primer, UMI, and PCR primer are attached to the molecule template in DADA2? I know DADA2 can handle PacBio sequences, but I m not sure in cases where we add UMI-barcode to the molecule. Thanks for your help

**Comment by benjjneb** [comment]
DADA2 does not currently handle UMI sequences, i.e. it will not create consensuses out of molecules sharing a UMI as is desired when using UMI approaches.

**Comment by joey711** [comment]
@Ellie-BM  I've been wanting to see data like this for a long time now, specifically knowing about the use-case in HIV that predates dada2. I wanted to basically "steal" this method from HIV for microbiome amplicon-seq. I still think this is probably where the field needs to go... In principle I think the UMI and biological sequence regions could be denoised separately, and then linked back together informatically via read identifiers. The UMI is obviously separate information about the provenance of observed sequences, would allow for correction of that pesky amplification bias, etc...

**Comment by Ellie-BM** [comment]
@joey711 True, UMIs are useful for decreasing error rates. But still the error rate is substantial which can happens in the insert, which means we have a bunch of reads with the same UMI, do not have the same insert. How do we make a consensus there? Sometimes the inserts are different but pretty close, sometimes they are very different and that is a problem. Errors don’t just happen in the insert they can happen in the UMI. Sometimes there are errors in UMI, i.e., there should be identical while they are not and it can lead to create a fake apparent new molecule and that is another issue.   I used dada2 for microbiome amplicon-seq before and I was quite content with it.  That is why I am curious to see if DADA2 can handle this type of data. Most of the available methods are not sophisticated enough, at least the ones I read so far. I would appreciate if you have any suggestion on this. :)

**Comment by Ellie-BM** [comment]
@benjjneb Just curious if you have thought about or suggest a useful method to handle this type of data with these kind of error? :)

**Comment by joey711** [comment]
@Ellie-BM I understand those issues with interpretation. I would be curious to see how often these different contradictions occur in real data after both the UMI and the insert have been denoised (separately) with a method like dada2. Since the two sequence regions (UMI, insert) are generated with separate processes and are distributed quite differently, I can't imagine not denoising them separately. At which point we would have to merge (as I suggested above) and then could then look at examples of these contradictions and summarize how it seems to pan out. My hypothesis is that a lot of the "insert differs from other members of the same UMI group" are errors that would have been caught by dada2, anyway. With possible exception of rare UMI collisions or early-cycle PCR errors, that latter being something we know dada2 can't catch but UMI would. On the other hand, the PCR-bias correction from the UMI is not possible without it. Between that and the overall improvement to accuracy lent by an additional independent source of information seems pretty compelling to me. At first glance I don't think you need another tool, you need a special execution of dada2 on the UMI regions, then the insert, then remerging. Which isn't off-the-shelf in dada2. If an example approach exists somewhere else we could look at that, but it sounds like you haven't found one.

**Comment by Ellie-BM** [comment]
@joey711  I am pretty new on UMIs and SMRT sequencing method. So my apologies if I am not very clear or my understanding is still limited about some of the terms. I know an example that they use phylogenetic tree and see how closely insert sequences are and how closely UMIs are by simply putting a UMI bases as suffix name for that insert in the tree. So when they have inserts that are identical and UMIs that are very close, that is a sign of error in UMI that created the fake bin or fake molecule, so they would collapse the UMIs with edit distance 1.  For non identical inserts and UMIs bin with edit distance 1 they look into high/low patterns and define a cutoff to remove that UMI bin. This is very biased approach. 

I think what you suggest is something I really like to apply on this HIV and then Covid-19 data for preprocessing and see what the results are and I can tell you how good the method works on these data. Just to clarify here, you suggest, I denoise UMIs which each UMI has 8 bases using DADA2 and do the same thing on insert sequences, and put them back together and see how the results will look like compared to other methods? Am I correct?

**Comment by joey711** [comment]
@Ellie-BM yes that sounds like what I'm suggesting. 8 bases is not a lot, so I wonder if we'll get into trouble there somehow. The good news is that the computation speed should be fast :-) The joining will be a bit tricky because you'll probably need to do it via sequence identifier. I have some related code laying around somewhere... Let me know if I can help

**Comment by Ellie-BM** [comment]
@joey711 Great! I ll work on this and let you know the result as soon as I get and yes I also think I need to use sequence identifier. I will let you know if I need any help on this, Thanks so much. Also,  is it possible to share the code you mentioned above with me when you get a chance :-)? I am sure it will be helpful. Thanks so much

**Comment by joey711** [comment]
@benjjneb please jump in if I've missed something quite boneheaded here. I'm pretty sure that we talked about this idea a long while back and I don't remember encountering a "gotcha" back then. Hope it isn't my rosy memory getting the better of me. We would have done this at Pendulum by now except that the improvement in accuracy was difficult to justify the technical development cycles to spin up a new method... The 8-base UMI region actually sounds ideal in some respects, because it should be relatively uniformly sampled from that 4^8 sequence space. The method implies that you're sampling a much smaller number of molecules than 4^8, and it is those collisions or near-collisions that would be an interpretive problem (along with any practical issues with the current implementation).

**Comment by joey711** [comment]
The code I mentioned was from an early draft for using read identifiers during post-denoising interpretation, so I would have to adapt a bit for this purpose and check that it still matches up (or tweak). If you send me some small anonymized otherwise unuseful test data I can adapt it and make a MRE for you to follow.

**Comment by Ellie-BM** [comment]
@joey711 Hi Joey, what sort of data would be useful for you for adapting read identifiers?  Raw data before doing CSS or CSS data?

**Comment by joey711** [comment]
Oh, the second one, I think. The subreads of the CCS are not going to be particularly useful to dada2, afaik. So, ideally you could share smallish number of CCS that include the UMI. Can you share as fastq? And assuming fastq, I also need to know the regex that explains the sequence header (aka read ID), and other details about the construct so that I know how to split apart the UMI segment from the biological sequence (aka insert). Alternatively these could be separate files if that was easier (one file for UMI-only, one file for insert-only), as long as matching them up later by ID was still possible. Does that make sense? Is this already close at hand? The easiest option might be a technical control sample so that you could share without needing to do any complicated filtering/modification.

**Comment by benjjneb** [comment]
@Ellie-BM Sorry I don't have any concrete other software to suggest, largely because I haven't had the chance to work with long-read UMI data myself yet. You may want to take a look at this preprint though: https://www.biorxiv.org/content/10.1101/645903v3.abstract

They are using UMIs with PacBio and ONT data and reporting some very impressive accuracy results in the UMI-consensus sequences.

**Comment by Ellie-BM** [comment]
@joey711 Hi Joey, I just sent the test data to your email account listed in GitHub.Please let me know if you have any questions. Thank you for your help on this. I appreciate it.

**Comment by Ellie-BM** [comment]
@benjjneb Thank you very much for the paper suggestion. I ll look into it.

**Comment by Ellie-BM** [comment]
@joey711 Hi Joey, I just sent you an email related to data from my work email, in case it went astray.  Thanks
